### PR TITLE
revisit-pagination-in-ts

### PR DESCRIPTION
**Pull request by qiaozha**
revisit pagination guideline in typescript.

### CODE REVIEW

**Review by joheredi — DISMISSED** [review]

**Review by xirzec — APPROVED** [review]
Changes seem righteous to me, though I'd appreciate if @deyaaeldeen could take a quick look to ensure we didn't miss any subtle nuance of the pagination contract.

File: docs/typescript/design.md [hunk 1]
```diff
@@ -491,7 +487,10 @@ for await (const item of client.listItems()) {
     console.log(item);
 }
 
-for await (const page of client.listItems().byPage({ maxPageSize: 50 })) {
+// usage of continuationToken with byPage
+const previousPage = client.listItems().byPage().next();
```
Comment (deyaaeldeen) [inline docs/typescript/design.md:491]:
```suggestion
const previousPage = await client.listItems().byPage().next();
```

File: docs/typescript/design.md [hunk 2]
```diff
@@ -500,45 +499,57 @@ interface Item {
     name: string;
 }
 
-interface Page {
-    continuationToken: string;
-    items: Item[];
+type ContinuablePage<TElement, TPage = TElement[]> = TPage & {
+  /**
+   * The token that keeps track of where to continue the iterator
```
Comment (deyaaeldeen) [inline docs/typescript/design.md:504]: nitpick:
```suggestion
   * A token that identifies a certain page.
```

File: docs/typescript/design.md [hunk 3]
```diff
@@ -500,45 +499,57 @@ interface Item {
...
+   */
+  continuationToken?: string;
+};
+/**
+ * An interface that allows async iterable iteration both to completion and by page.
```
Comment (deyaaeldeen) [inline docs/typescript/design.md:509]: nitpick:
```suggestion
 * An iterator that also enables iteration over both items and pages of items.
```

File: docs/typescript/design.md [hunk 4]
```diff
@@ -500,45 +499,57 @@ interface Item {
...
+  /**
+   * The next method, part of the iteration protocol
+   */
+  next(): Promise<IteratorResult<TElement>>;
+  /**
+   * The connection to the async iterator, part of the iteration protocol
+   */
+  [Symbol.asyncIterator](): PagedAsyncIterableIterator<TElement, TPage, TPageSettings>;
+  /**
+   * Return an AsyncIterableIterator that works a page at a time
```
Comment (deyaaeldeen) [inline docs/typescript/design.md:525]:
```suggestion
   * Return an iterator over pages of items.
```

File: docs/typescript/design.md [hunk 5]
```diff
@@ -500,45 +499,57 @@ interface Item {
...
+      return buildPagedAsyncIterator(
+        context,
+        () =>
+          _listItemsSend(
+            context,
+            options,
+          ),
+        _listItemsDeserialize,
+        { itemName: "value", nextLinkName: "nextLink" },
+      );
```
Comment (deyaaeldeen) [inline docs/typescript/design.md:542]: If we're going to show this code, I recommend giving context on all referenced definitions including `buildPagedAsyncIterator`.

Comment (qiaozha) [reply docs/typescript/design.md]: I have removed this part, not sure but still kept the interface PagedAsyncIterableIterator definition part, not sure if you have concerns about those?

**Review by deyaaeldeen — APPROVED** [review]

**Pull request merged**